src/d3d9/d3d9_device.cpp

```diff
@@ -4702,6 +4702,27 @@ namespace dxvk {
     if (unlikely(!desc.IsLockable))
       return D3DERR_INVALIDCALL;
 
+    D3DRESOURCETYPE type = pResource->GetType();
+
+    // Volume textures in D3DPOOL_DEFAULT must have the D3DUSAGE_DYNAMIC flag set in order to be lockable.
+    if (unlikely(type == D3DRTYPE_VOLUMETEXTURE && desc.Pool == D3DPOOL_DEFAULT && !(desc.Usage & D3DUSAGE_DYNAMIC)))
+      return D3DERR_INVALIDCALL;
+
+    if (unlikely(pBox != nullptr)) {
+      D3D9_FORMAT_BLOCK_SIZE blockSize = GetFormatBlockSize(desc.Format);
+
+      // LockImage calls on D3DPOOL_DEFAULT surfaces and volume textures with formats
+      // which need to be block aligned, must be validated for mip level 0.
+      if (MipLevel == 0 && (type == D3DRTYPE_VOLUMETEXTURE ||
+        (type != D3DRTYPE_VOLUMETEXTURE && desc.Pool == D3DPOOL_DEFAULT))
+        && blockSize.Width > 0 && blockSize.Height > 0
+        && ((pBox->Left   && (pBox->Left   & (blockSize.Width  - 1))) ||
+            (pBox->Top    && (pBox->Top    & (blockSize.Height - 1))) ||
+            (pBox->Right  && (pBox->Right  & (blockSize.Width  - 1))) ||
+            (pBox->Bottom && (pBox->Bottom & (blockSize.Height - 1)))))
```

Comment:
This isn't easy to read. Please move some of those conditions into separate booleans.

`)))))` is a bit of a red flag.

Reply:
Done.